src/index.ts

```diff
@@ -68,5 +68,7 @@ function removeNullOrUndef(params: ParamMap) {
 }
 
 function notNullOrUndefined(v: any) {
+  if (v === null || v === undefined || typeof v === "string" && v.trim().length <= 0) throw new TypeError(`Path parameter ${v} is not Valid`);
```

Comment:
This will throw for null or undefined query params too which are valid (and the reason this function exists). I would solve this in a different way.

Reply:
@balazsbotond ohh,let me know once you are done,I would like to learn 👍

Reply:
@balazsbotond  can you discuss here?

Reply:
I think we misunderstood each other :) I just wanted to five you a hint that this needs to be solved in a different way. You are still welcome to do it!

Reply:
I think a simple way to do it would be to call cleanParams *after* the call to path in urlcatImpl, so your code from the previous PR would handle the null and undefined cases.

Reply:
@balazsbotond ohh for that we don't even need to set new error message.right!!

Reply:
> 
> 
> I think a simple way to do it would be to call cleanParams _after_ the call to path in urlcatImpl, so your code from the previous PR would handle the null and undefined cases.

I have tried this,for this case there is no need for `removeNullOrUndef` function at all ,because null and undefined are handled in allowedTypes. What you say? And what about missing value like blank string?

Reply:
You are right about empty strings - I think we shouldn't allow them either in paths.

In query params though, null and undefined are ignored by design, and empty strings are also allowed by design (an empty query parameter is perfectly valid). So I think we should still call `removeNullOrUndef` but only after the call to `path`.

Reply:
@balazsbotond like this?
`function urlcatImpl(pathTemplate: string, params: ParamMap, baseUrl?: string) {`
 ` const { renderedPath, remainingParams } = path(pathTemplate, params);`
  `const cleanParams = removeNullOrUndef(remainingParams);`
  `const renderedQuery = query(cleanParams);`
  `const pathAndQuery = join(renderedPath, '?', renderedQuery);`
  `return baseUrl`
  `  ? join(baseUrl, '/', pathAndQuery)`
   ` : pathAndQuery;`
`}`

Reply:
Yes, exactly. And then inside `path`, also check for an empty string, as you have suggested.

Btw, you can format code blocks in comments using triple backticks:

````
```ts
function doSomething() {
  // do it
}
```
````

Result:

```ts
function doSomething() {
  // do it
}
```

This preserves indentation, uses syntax highlighting (in this case, for ts), and you don't have to put each line between backticks.

Reply:
@balazsbotond Hahahah nice thanks!!

Reply:
@balazsbotond Please check #24.I have made new pr because old pr would have come with conflict.
and
```
it('Can handle allowed Type String as param', () => {
    const expected = 'http://example.com/path/%20test';
    const actual = urlcat('http://example.com/path/:p', { p: ' test' });
    expect(actual).to.equal(expected);
  });
```

is this correct? or you want to sanitize params?

Reply:
This is correct. I think we should follow the principle of least surprise here and not sanitize the input - it might cause unexpected behavior for users.

Reply:
If you merge the master branch into your feature branch, resolve conflicts locally and push them, you can still use this PR. I wouldn't want to close this one because of the comment history. Could you please do this and delete the new PR (#24)?

Reply:
> 
> 
> If you merge the master branch into your feature branch, resolve conflicts locally and push them, you can still use this PR. I wouldn't want to close this one because of the comment history. Could you please do this and delete the new PR (#24)?

Yes  Yes I will do it

Reply:
> 
> 
> This is correct. I think we should follow the principle of least surprise here and not sanitize the input - it might cause unexpected behavior for users.

what is least surprise?

Reply:
It is a commonly used software design principle that says that software should, if possible, work the way you expect it. See: https://en.wikipedia.org/wiki/Principle_of_least_astonishment

Reply:
> > If you merge the master branch into your feature branch, resolve conflicts locally and push them, you can still use this PR. I wouldn't want to close this one because of the comment history. Could you please do this and delete the new PR (#24)?
> 
> Yes Yes I will do it

Thank you!

Reply:
You can also resolve conflicts using the web editor:

![merge](https://user-images.githubusercontent.com/2791280/94674332-4ebc6500-0318-11eb-986f-d150670dcb5f.png)

It's much easier this way.

Reply:
@balazsbotond Please give me some time, I will do it in few hours